test/files/run/t10551.scala

```diff
@@ -0,0 +1,48 @@
+package test {
+  object NotNoPrefix {
+    final class Id[A](val a: A) extends AnyVal
+    final class Ids[A](val as: Seq[A]) extends AnyVal
+    final class Bid[A, B](val ab: Map[A, B]) extends AnyVal
+  }
+}
+
+object Test extends App {
+  import test.NotNoPrefix._
+
+  println(classOf[Id[Int]])
+  println(classOf[Id[_]])
+
+  println(classOf[Ids[Int]])
+  println(classOf[Ids[_]])
+
+  println(classOf[Bid[Int, Int]])
+  println(classOf[Bid[Int, _]])
+  println(classOf[Bid[_, Int]])
+  println(classOf[Bid[_, _]])
+
+  type Iddy[A] = Id[A]
+  type Idsy[A] = Ids[A]
+  type Biddy[A, B] = Bid[A, B]
+  type Biddouble[A] = Bid[A, Double]
+  type Bixt[L] = Biddouble[_]
+  type Bixty = Bixt[_]
+
+  println(classOf[Iddy[Int]])
+  println(classOf[Iddy[_]])
+
+  println(classOf[Idsy[Int]])
+  println(classOf[Idsy[_]])
+
+  println(classOf[Biddy[Int, Int]])
+  println(classOf[Biddy[Int, _]])
+  println(classOf[Biddy[_, Int]])
+  println(classOf[Biddy[_, _]])
+
+  println(classOf[Biddouble[Int]])
+  println(classOf[Biddouble[_]])
+
+  println(classOf[Bixt[Int]])
+  println(classOf[Bixt[_]])
+
+  println(classOf[Bixty])
+}
```

Comment:
Old vs new:

```diff
--- /tmp/old    2018-01-12 15:59:36.000000000 +1000
+++ /tmp/new    2018-01-12 16:00:36.000000000 +1000
@@ -1,21 +1,21 @@
 class test.NotNoPrefix$Id
-class java.lang.Object
+class test.NotNoPrefix$Id
+class test.NotNoPrefix$Ids
 class test.NotNoPrefix$Ids
-interface scala.collection.Seq
 class test.NotNoPrefix$Bid
-interface scala.collection.immutable.Map
-interface scala.collection.immutable.Map
-interface scala.collection.immutable.Map
+class test.NotNoPrefix$Bid
+class test.NotNoPrefix$Bid
+class test.NotNoPrefix$Bid
+class test.NotNoPrefix$Id
 class test.NotNoPrefix$Id
-class java.lang.Object
 class test.NotNoPrefix$Ids
-interface scala.collection.Seq
+class test.NotNoPrefix$Ids
+class test.NotNoPrefix$Bid
+class test.NotNoPrefix$Bid
+class test.NotNoPrefix$Bid
+class test.NotNoPrefix$Bid
+class test.NotNoPrefix$Bid
+class test.NotNoPrefix$Bid
+class test.NotNoPrefix$Bid
+class test.NotNoPrefix$Bid
 class test.NotNoPrefix$Bid
-interface scala.collection.immutable.Map
-interface scala.collection.immutable.Map
-interface scala.collection.immutable.Map
-class test.NotNoPrefix$Bid
-interface scala.collection.immutable.Map
-interface scala.collection.immutable.Map
-interface scala.collection.immutable.Map
-interface scala.collection.immutable.Map
```

Reply:
I manually checked that the new, correct output also holds for `classTag`, rather than `classOf`.

Reply:
That suggests that [class tag materialization](https://github.com/scala/scala/blob/71e76e8c056f0dad2447356847433ad33f3321a6/src/compiler/scala/reflect/reify/package.scala#L63-L71) was already doing the right thing, as I don't think it calls into code paths changed here.

I think we should at least expand this test case to assert identical results with `classTag`/`classOf`, and ideally create a link between the different parts of the code by changing `tpe.dealiasWiden.erasure` in `reifyRuntimeClass` to use the new method, `applyInArray`.

Reply:
Here's the expanded test: https://github.com/scala/scala/commit/da003548092c10f03f985ed6fbde1db300a4e76d

I tried unifying the implementation of `reifyRuntimeClass` with your code and it eluded me. I'm happy to leave that work aside and merge this PR once you add my commit with the class tag tests.

Reply:
So I did a bit of poking at that, too. I don't think we actually need to talk about erasures in `reifyRuntimeClass` at all; this seems to work perfectly well:

```diff
@@ -65,9 +68,7 @@ package object reify {
         val componentErasure = reifyRuntimeClass(global)(typer0, componentTpe, concrete)
         gen.mkMethodCall(currentRun.runDefinitions.arrayClassMethod, List(componentErasure))
       case _ =>
-        var erasure = tpe.erasure
-        if (tpe.typeSymbol.isDerivedValueClass && global.phase.id < global.currentRun.erasurePhase.id) erasure = tpe
-        gen.mkNullaryCall(currentRun.runDefinitions.Predef_classOf, List(erasure))
+        gen.mkClassOf(tpe)
     }
   }
```

Generating the call to `classOf` requires calculating the erasure, since `classOf`'s type argument isn't poly-kinded (so we need to make sure we pass it a `*`-kinded type) and since `typedClassOf` will only accept class types. However, all that `typedClassOf` does (if it succeeds) is replace the `TypeApply` with a `TypeTree` wrapping the argument, so we can just make the `TypeTree` here!

The generated `TypeTree` will be subject to the same treatment in `erasure` as if the user had written `ClassTag[T](classOf[T])` directly.

I'll push that up as a separate commit and squash if it makes sense.

Reply:
I also want to consider renaming `applyInArray`; it's not the most explanatory in some of it's new usages. As I said, though, I can't think of something terse and better.